Add Cardinal to SALAMANDER

cticenhour opened the pull request:
Closes #38 

@meltawila As discussed earlier, please build this version of SALAMANDER and run it against your Cardinal inputs from #30 (and any other inputs you are working on, as you like). Please also add any comments you have on build instructions and usage issues to this thread so that they can be addressed as part of the review process. Thanks! 

Requesting @simopier for the formal review once this test run is complete and any comments from that test are fully addressed.

cticenhour committed:
Add Cardinal submodule

Refs #38

cticenhour committed:
Add Cardinal to Makefile

This enables linking to the executable, but object registration is not done yet

cticenhour committed:
Update Cardinal submodule

cticenhour committed:
Get Cardinal and contribs (OpenMC, DAGMC, MOAB) fully linking to FENIX

Still need to somewhat automate build process for FENIX end users (via a script).

cticenhour committed:
Add OpenMC download location in FENIX to gitignore file

cticenhour committed:
Add initial cardinal build script

cticenhour committed:
Add Cardinal to installation instructions

Refs #38

cticenhour commented on doc/content/getting_started/installation.md:
````diff
@@ -59,20 +59,43 @@ To compile FENIX, first make sure that the conda MOOSE environment is activated
+and subsequently set the location of those cross sections in your environment:
+
+```bash
+export OPENMC_CROSS_SECTIONS=~/projects/FENIX/cross_sections/cross_sections.xml
````
Reminder to add a warning to the documentation here letting the user know that this must be set in the environment *every time* the Terminal is reloaded (unless it is set persistently in the shell profile).

simopier replied:
> Reminder to add a warning to the documentation here letting the user know that this must be set in the environment every time the Terminal is reloaded (unless it is set persistently in the shell profile).

Are you going to add this reminder in this PR?

cticenhour replied:
Yes - I like to avoid too many changes after I've requested reviews, and then integrate my comments with the review changes. :)

cticenhour replied:
Addressed in 2927d33

cticenhour commented:
Note that some changes to the recipes need to occur before this can be merged. I am working on a recipe PR that I will get reviewed and merged next week.

cticenhour commented:
All - also happy to take feedback on the script configuration currently used here. An alternative is doing everything in the Makefile, which should remove at least one step from the process as currently outlined, and require fewer recipe changes.

meltawila commented on scripts/update_and_rebuild_cardinal.sh:
```diff
@@ -0,0 +1,139 @@
+if [ -n "$HDF5_DIR" ]; then
+  export HDF5_ROOT=$HDF5_DIR
+else
+  echo -e "${RED}ERROR: HDF5 not found; please set HDF5_DIR in your environment!${NC}"
```
I think this is never set automatically via conda (at least on my system). `export HDF5_ROOT=$CONDA_PREFIX` works for me

meltawila replied:
I also used `export LD_LIBRARY_PATH=$CONDA_PREFIX/lib:$LD_LIBRARY_PATH` which I know cause errors sometimes when it's not set properly

meltawila commented:
Thank you Casey! I was able to follow your instructions and install with Cardinal enabled. I ran Cardinal test suite as well and got 384 passed, 254 skipped, 0 pending, 6 FAILED. The failed tests are probably out of the scope of this PR (i.e. one of them uses THM which we're not including here). otherwise it looks great to me!

simopier commented on doc/content/getting_started/installation.md:
````diff
@@ -59,20 +59,43 @@ To compile FENIX, first make sure that the conda MOOSE environment is activated
+the build script from the FENIX clone directory:
+
+```bash
+scripts/update_and_rebuild_cardinal.sh
````
```suggestion
cd ~/projects/FENIX
scripts/update_and_rebuild_cardinal.sh
```

cticenhour replied:
Addressed in 2927d33

simopier reviewed (changes requested):
Looks good to me. Thank you @cticenhour for adding this!
I'll re-review after the new recipe is added.

simopier commented:
Converting as draft until we get the new recipe merged.

simopier commented:
I think it would be nice to get @loganharbour's review as well.

cticenhour commented:
> The failed tests are probably out of the scope of this PR (i.e. one of them uses THM which we're not including here).

@meltawila Could also be related to our MOOSE submodule being ahead of Cardinal's; that would be my guess with only a few tests failing. 

Thanks for checking with your workflow!

cticenhour committed:
Address initial review comments

cticenhour committed:
Add new content to documentation

- Adds Cardinal content
  - Excludes Nek-related content
  - Adds landing page for BulkEnergyConservationICAction
  - Adds landing page for VolumetricHeatSourceICAction
  - Removes excess syntax information for both IC actions
  - Adds landing page for Cardinal/ICs custom syntax
- Adds module content
- Add SQA dependencies and SQA requirement content for modules and apps

Refs #38 and #52

cticenhour commented:
With the most-recent documentation commit, a small adjustment to TMAP8 is now required. idaholab/TMAP8#153 is pending review and merge.

simopier commented on doc/config.yml:
```diff
@@ -59,26 +89,51 @@ Extensions:
+        categories: # commented-out categories are not available
             fenix: !include ${ROOT_DIR}/doc/sqa_fenix.yml
             framework: !include ${MOOSE_DIR}/framework/doc/sqa_framework.yml
-            ray_tracing: !include ${MOOSE_DIR}/modules/ray_tracing/doc/sqa_ray_tracing.yml
```
Why are you removing ray tracing?

simopier replied:
Oh sorry, I just saw that it was there twice. Good catch!

cticenhour replied:
Alphabetical module listing is our friend 😄

cticenhour replied:
It wasn't there twice - I just moved it.

simopier commented on doc/config.yml:
```diff
@@ -59,26 +89,51 @@ Extensions:
             stochastic_tools: !include ${MOOSE_DIR}/modules/stochastic_tools/doc/sqa_stochastic_tools.yml
+            thermal_hydraulics: !include ${MOOSE_DIR}/modules/thermal_hydraulics/doc/sqa_thermal_hydraulics.yml
+            tmap8: !include ${ROOT_DIR}/tmap8/doc/sqa_tmap8.yml
+            #cardinal: !include ${ROOT_DIR}/cardinal/doc/sqa_cardinal.yml
```
Why is Cardinal commented out?

cticenhour replied:
Cardinal doesn't use the MOOSE SQA system.

cticenhour replied:
I also say this in a comment on Line 109

simopier commented on doc/sqa_fenix.yml:
```diff
@@ -2,6 +2,23 @@ directories:
+    - stochastic_tools
+    - thermal_hydraulics
+    - tmap8
+    #- cardinal
```
Again, I'm wondering why cardinal is commented out.

cticenhour replied:
Again, Cardinal doesn't use the SQA system. This is there as a placeholder if it does (and a reminder to devs that we didn't neglect it). I say this as a comment on Line 5.

simopier reviewed (changes requested):
Just a couple questions.

cticenhour commented:
Based on a discussion today with @loganharbour, I think I will adjust this script slightly and the `Makefile` in order to facilitate the running of the script as part of the normal `make` command. This would bring the build process back to something more in-line with a "normal" MOOSE application and ease potential build complications for users.

cticenhour committed:
Transition to running the cardinal build in the prebuild Makefile rule

cticenhour commented:
One last adjustment to MOOSE was required, which has now been merged. Once that makes it to the MOOSE submodule, I think we'll be ready to go.

moosebuild commented:
All [jobs](https://civet.inl.gov/event/187693/) on d887e3b : invalidated by @GiudGiud

post-submodule update

GiudGiud commented:
I think the desired submodule update was not done yet

moosebuild commented:
All [jobs](https://civet.inl.gov/event/187693/) on d887e3b : invalidated by @simopier

submodule update

cticenhour committed:
Fixup build Cardinal script for case when PETSc is pre-installed and HDF5 needs to be found in PETSc location

cticenhour commented on doc/content/getting_started/installation.md:
````diff
@@ -59,20 +59,50 @@ To compile FENIX, first make sure that the conda MOOSE environment is activated
 ```bash
 cd ~/projects/FENIX
 git submodule update --init moose
+git submodule update --init tmap8
````
```suggestion
git submodule update --init tmap8
git submodule update --init cardinal
```

cticenhour commented on doc/content/getting_started/installation.md:
````diff
@@ -59,20 +59,50 @@ To compile FENIX, first make sure that the conda MOOSE environment is activated
+```bash
+cd ~/projects/FENIX
+scripts/update_and_rebuild_cardinal.sh
+```
````
```suggestion
```

Cardinal is now built within the Makefile rules rather then being done manually.

cticenhour committed:
Fixup installation documentation given Makefile changes

simopier commented on doc/content/getting_started/installation.md:
````diff
@@ -59,20 +59,44 @@ To compile FENIX, first make sure that the conda MOOSE environment is activated
 ```
 
-Then navigate to the FENIX clone directory and download the MOOSE submodule:
+Then navigate to the FENIX clone directory and download the MOOSE and TMAP8 submodules:
````
```suggestion
Then navigate to the FENIX clone directory and download the MOOSE, TMAP8, and Cardinal submodules:
```

simopier reviewed (changes requested):
Just a small documentation fix given the latest update.

cticenhour committed:
Address review comment

Co-authored-by: Pierre-Clement Simon <pierreclement.simon@gmail.com>

cticenhour committed:
Move submodule updates to getting started instructions

moosebuild commented:
All [jobs](https://civet.inl.gov/event/191110/) on d1b0527 : invalidated by @cticenhour

Invalidating due to recipe update (submodules should be fetched properly now)

moosebuild commented:
All [jobs](https://civet.inl.gov/event/191110/) on d1b0527 : invalidated by @cticenhour

Fix MOAB typo in recipe

cticenhour commented:
The TMAP8 portion of this PR has been superceded by #65, so the TMAP8 components will need to be removed. Differing errors are occurring during the Cardinal-inclusive build on Linux CI and on a local container duplicating the CI environment. I will tackle this when back from England.

cticenhour committed:
Building Cardinal contribs directly in the FENIX makefile

This borrows heavily from the workflow in "cardinal/Makefile" and uses "cardinal/config" *.mk files to do the contrib builds and linking directly in the FENIX makefile, generating a single executable

cticenhour committed:
Update cardinal submodule

cticenhour committed:
Fixup breakages with updated cardinal

- Generate CardinalRevision.h due to installation changes, turn on FenixRevision.h as well (not required yet)
- Add dummy redirect file for AddTallyAction for use with SQA

moosebuild commented:
All [jobs](https://civet.inl.gov/event/192418/) on 3abe93c : invalidated by @cticenhour

Recipe changes - Cardinal contrib submodules available

cticenhour committed:
Adjust HDF5_DIR in the case that PETSc has built HDF5

This will hopefully robustify this for the case when PETSc is installed in a container as well as locally (within an arch-* location)

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2415829/) on 8473f78 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/fenix/docs/PRs/53/site)

This comment will be updated on new commits.

cticenhour committed:
Add cardinal and dependencies to FENIX unit Makefile

cticenhour committed:
Add apptainer context for generating an apptainer image of FENIX that includes Cardinal contribs and cross sections

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2415833/) on 8473f78 wanted to post the following:


## Coverage

No coverage report(s) were found for the base commit 8dfa4a76b5.
[Full coverage report](https://mooseframework.inl.gov/fenix/docs/PRs/53/coverage)



This comment will be updated on new commits.

cticenhour commented:
The remaining failure is related to compiler warnings coming out of Cardinal/OpenMC. I intend to fix these up (or set them up to be ignored in the case of OpenMC), but that will require 1 libMesh PR, 1 Cardinal PR, and then a follow-on PR here to update Cardinal. 

So as not to hold this up any further, I am simply turning off `-Werror` checking for the debug PR/devel recipe in idaholab/civet_recipes#1841. I have made an issue (#68) that this is a problem that should be resolved per our code standards and code cleanliness guidelines.

I will work on fixing this up as soon as I am able, but it might have to wait until the next FY.

cticenhour committed:
Add tip in documentation about speeding up Cardinal contrib builds

simopier commented on doc/content/getting_started/installation.md:
```diff
@@ -85,12 +115,25 @@ build the code and run the tests. The number in that flag can be changed to the
+FENIX (and its underlying dependencies MOOSE, TMAP8, and Cardinal) is under heavy development and is
 updated on a continuous basis. Therefore, it is important that the local copy of FENIX be periodically
 updated to obtain new capabilities, improvements, and bugfixes. Weekly updates are recommended as,
 at minimum, the MOOSE submodule within FENIX is updated up to several times a week.
```
I cannot point to it or make a suggestion there directly, but the instructions to update FENIX should also be updated. 
It currently only update moose, and should update TMAP8 and Cardinal with something  

```
git submodule update moose
git submodule update tmap8
git submodule update cardinal
```
Note that this change should occur in two different places.

cticenhour replied:
Addressed in ffa88f9.

simopier reviewed (changes requested):
Just a minor request, all good after that!

cticenhour committed:
Address review comments

simopier reviewed (dismissed):
Thank you!

loganharbour reviewed (commented):
You want cardinal to be a _requirement_ of fenix?

cticenhour commented:
@loganharbour This was the intent yes. If this ends up being onerous or disruptive from a CI perspective, we can always make it optional. Is this what you are wary of?

@simopier do you have a preference based on Logan's question?

loganharbour commented:
No CI concerns. User concerns. In almost all cases where we have coupling, it's optional. If the application has features that do not depend on the coupling application, it should be able to be optional. 

Take a moose dev who is trying to make a patch. IMO don't make them compile cardinal unless they need to.

cticenhour commented:
Sounds like we're measuring by degrees here. We very often make folks compile multiple MOOSE-based submodules for an application which are "on" at all times. Cardinal's relative weight is the defining difference, and I don't think it's that onerous since we're only adding OpenMC and DAGMC only - nekRS is explicitly left out. 

That being said, I don't have a strong preference; I just wanted to make sure the build and setup worked smoothly for users and devs. The Makefile and source changes to make it optional are not hard to do. As PI, I want to give @simopier the final say.

loganharbour commented:
I feel pretty strongly about it. This just isn't how we do it it. Not being able to compile it on its own removes semblance of what product this is. And in the future when we make dynamic loading the path forward, you'll have to separate them. Potentially allowing things to mix in to the code base that are cardinal dependent.

meltawila reviewed (approved):
Should we update `doc/content/getting_started/hpc` as well? 
Mainly for adding instructions to build PETSc, libMesh, and WASP as I think it'd be required now. I used:
```
./moose/scripts/update_and_rebuild_petsc.sh
./moose/scripts/update_and_rebuild_libmesh.sh
./moose/scripts/update_and_rebuild_wasp.sh
```

cticenhour commented:
@meltawila ~~good call - the HPC instructions need to be updated a bit.~~ EDIT: Actually, we do `include` common directions between files. They are updated automatically with these changes.

However, I don't think we need to add the update and rebuild instructions. They should be provided with the latest `moose-dev` module. Has this not been your experience?

cticenhour committed:
Make Cardinal an optional dependency of FENIX

simopier commented on Makefile:
```diff
@@ -42,82 +43,16 @@ else
-export HDF5_ROOT
+# Check for optional dependencies and, if found, configure for building.
+# Set default for ENABLE_CARDINAL.
+ENABLE_CARDINAL := no
```
Doesn't this define the default as NOT having CARDINAL? 

I think the default should be to include cardinal.

cticenhour replied:
Addressed in 93cb186.

cticenhour committed:
Address review comment

Default for Cardinal is changed to 'yes' instead of 'no'

simopier commented on Makefile:
```diff
@@ -24,8 +43,14 @@ else
+# Check for optional dependencies and, if found, configure for building.
+include config/check_deps.mk
+
+ifeq ($(ENABLE_CARDINAL),yes)
```
```suggestion
# ENABLE_CARDINAL = yes by default, but is set to "no" automatically if Cardinal is not found via CARDINAL_DIR. 
ifeq ($(ENABLE_CARDINAL),yes)
```

simopier commented on config/check_deps.mk:
```diff
@@ -0,0 +1,18 @@
+define n
+
+
+endef
```
What is that for?

cticenhour replied:
It's for spacing and newlines in the printed output of the warning and info messages later in this document. See the references to `$n` there.

simopier commented on config/check_deps.mk:
```diff
@@ -0,0 +1,18 @@
+CARDINAL_CONTENT     := $(shell ls $(CARDINAL_DIR) 2> /dev/null)
+
+ifeq ($(CARDINAL_CONTENT),)
+  $(warning $n"Cardinal does not seem to be available. If usage of Cardinal is desired within FENIX, make sure that either the submodule is checked out$nor that CARDINAL_DIR points to a location with the Cardinal source.")
```
```suggestion
  $(warning $n"Cardinal does not seem to be available. If usage of Cardinal is desired within FENIX, make sure that either the submodule is checked out$nor that CARDINAL_DIR points to a location with the Cardinal source. $n In the meantime, FENIX will be built without Cardinal.")
```

simopier commented on unit/Makefile:
```diff
@@ -22,11 +37,20 @@ FRAMEWORK_DIR      ?= $(MOOSE_DIR)/framework
 # Use the TMAP8 submodule if it exists and TMAP8_DIR is not set
 # If it doesn't exist, and TMAP8_DIR is not set, then look for it adjacent to the application
-TMAP8_SUBMODULE    := $(CURDIR)/tmap8
+TMAP8_SUBMODULE    := $(CURDIR)/../tmap8
```
Great catch!

simopier commented on doc/content/getting_started/installation.md:
````diff
@@ -59,22 +59,52 @@ To compile FENIX, first make sure that the conda MOOSE environment is activated
 ```
 
-Then navigate to the FENIX clone directory and download the MOOSE and TMAP8 submodules:
+Then navigate to the FENIX clone directory and download the MOOSE, TMAP8, and Cardinal submodules:
````
The documentation needs to be updated to have a `Tips` message to explain how they can build FENIX without Cardinal if they desire.

simopier commented on doc/content/getting_started/installation.md:
```diff
@@ -85,12 +115,25 @@ build the code and run the tests. The number in that flag can be changed to the
 If FENIX is working correctly, all active tests will pass. This indicates that
 FENIX is ready to be used and further developed.
 
 ## Update FENIX
```
Just as above, this should be updated to tell users how to not use Cardinal.

simopier reviewed (changes requested):
Some small questions and a minor request for a documentation update to explain how to not build with Cardinal.

cticenhour committed:
Apply suggestions from code review

Co-authored-by: Pierre-Clement Simon <pierreclement.simon@gmail.com>

cticenhour committed:
Fixup intermittent documentation failure related to a vague link

cticenhour committed:
Add tip to installation instructions regarding disabling Cardinal

This is also reflected in the HPC instructions. Finally, improve spacing for warning message when Cardinal is disabled.

simopier commented on doc/content/getting_started/installation.md:
````diff
@@ -126,6 +182,12 @@ git checkout devel
+git submodule update --recursive contrib/DAGMC
+git submodule update contrib/moab
 ```
 
````
```suggestion

!alert! tip title=Disabling Cardinal
Cardinal is an optional dependency of FENIX and can be disabled to speed up build times, if Cardinal
capabilities are not needed. This can be done by simply not downloading the Cardinal submodule
(i.e., not performing the `git submodule update cardinal` command as well as the following Cardinal
dependency update commands above.)
!alert-end!

```

simopier reviewed (changes requested):
One last thing

cticenhour committed:
Apply suggestion from code review

Co-authored-by: Pierre-Clement Simon <pierreclement.simon@gmail.com>

simopier reviewed (approved):
Looks great to me, and this is a great contribution! Thank you @cticenhour!!

simopier commented:
@loganharbour, do you want to give this a another review before we merge it?

loganharbour reviewed (commented):
Looks good to me. We just need a no optional recipe

cticenhour commented:
@loganharbour I had already updated the follow-on recipe PR to include it if you want to do it there, or I can pull it out separately. 

EDIT: We're doing it separately, idaholab/civet_recipes#1874

loganharbour reviewed (approved):
Let's do it
